### PR TITLE
SONAR-6611 WS custom_measures/update update a custom measure

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/UpdateAction.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+      .setExampleValue("42")
+      .setRequired(true);
+
+    action.createParam(PARAM_VALUE)
+      .setRequired(true)
+      .setExampleValue("true")
+      .setDescription(measureValueDescription());
+
+    action.createParam(PARAM_DESCRIPTION)
+      .setRequired(true)
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/UpdateAction.java]: Is the description really required?

**Comment by jblievremont** [comment]
Not sure about the fact that custom measure description is mandatory, else LGTM!

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CustomMeasureValidator.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+@ServerSide
+public class CustomMeasureValidator {
+  private final TypeValidations typeValidations;
+
+  public CustomMeasureValidator(TypeValidations typeValidations) {
+    this.typeValidations = typeValidations;
+  }
+
+  public void setMeasureValue(CustomMeasureDto measure, String valueAsString, MetricDto metric) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CustomMeasureValidator.java]: MAJOR The Cyclomatic Complexity of this method "setMeasureValue" is 15 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CustomMeasureValueDescription.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      description.append("<li>");
+      description.append(String.format("%s - %s", metricType.description(), metricTypeWsDescription(metricType)));
+      description.append("</li>");
+    }
+    description.append("</ul>");
+
+    return description.toString();
+  }
+
+  private static String metricTypeWsDescription(Metric.ValueType metricType) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CustomMeasureValueDescription.java]: MAJOR Reduce the number of returns of this method 7, down to the maximum allowed 5. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "metricTypeWsDescription" is 19 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CustomMeasureJsonWriter.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    json.prop(FIELD_PROJECT_ID, component.uuid());
+    json.prop(FIELD_PROJECT_KEY, component.key());
+    json.prop(FIELD_DESCRIPTION, measure.getDescription());
+    json.prop(FIELD_VALUE, measureValue(measure, metric));
+    json.propDateTime(FIELD_CREATED_AT, new Date(measure.getCreatedAt()));
+    json.propDateTime(FIELD_UPDATED_AT, new Date(measure.getUpdatedAt()));
+    json.endObject();
+  }
+
+  private String measureValue(CustomMeasureDto measure, MetricDto metric) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CustomMeasureJsonWriter.java]: MAJOR The Cyclomatic Complexity of this method "measureValue" is 19 which is greater than 10 authorized. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CustomMeasureJsonWriter.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    Metric.ValueType metricType = Metric.ValueType.valueOf(metric.getValueType());
+    Double doubleValue = measure.getValue();
+    String stringValue = measure.getTextValue();
+
+    switch (metricType) {
+      case BOOL:
+        return doubleValue == 1.0d ? "true" : "false";
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/CustomMeasureJsonWriter.java]: CRITICAL Equality tests should not be made with floating point values. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1244)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* CRITICAL 1 critical
* MAJOR 7 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/core/measure/custom/db' package (org.codehaus.sonar:sonar-core:src/main/java/org/sonar/core/measure/custom/db) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/measure/custom/persistence' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/measure/custom/persistence) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/measure/custom/ws' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/measure/custom/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/UpdateAction.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    }
+  }
+
+  private void setValue(CustomMeasureDto customMeasure, @Nullable String value, MetricDto metric) {
+    if (value != null) {
+      validator.setMeasureValue(customMeasure, value, metric);
+    }
+  }
+
+  private void setDescription(CustomMeasureDto customMeasure, @Nullable String description) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/UpdateAction.java]: MAJOR Make "setDescription" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/UpdateAction.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    if (description != null) {
+      customMeasure.setDescription(description);
+    }
+  }
+
+  private void checkParameters(@Nullable String value, @Nullable String description) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/custom/ws/UpdateAction.java]: MAJOR Make "checkParameters" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 issues:
* CRITICAL 1 critical
* MAJOR 9 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/core/measure/custom/db' package (org.codehaus.sonar:sonar-core:src/main/java/org/sonar/core/measure/custom/db) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/measure/custom/persistence' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/measure/custom/persistence) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/server/measure/custom/ws' package (org.codehaus.sonar:sonar-server:src/main/java/org/sonar/server/measure/custom/ws) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

**Pull request merged**